File: src/main/java/systems/conduit/main/mixins/api/EntityMixin.java
package systems.conduit.main.mixins.api;

import net.minecraft.core.Direction;
import net.minecraft.core.Position;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.TextComponent;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.Pose;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.Vec3;
import net.minecraft.world.scores.Team;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
import systems.conduit.main.Conduit;
import systems.conduit.main.events.types.EntityEvents;

import java.util.*;
import java.util.stream.Collectors;

@Mixin(value = Entity.class, remap = false)
public abstract class EntityMixin implements systems.conduit.main.api.Entity {

    @Shadow public Level level;

    @Shadow public abstract UUID getUUID();
    @Shadow public abstract void sendMessage(Component component, UUID uuid);

    @Override
    public void sendMessage(String message) {
        sendMessage(new TextComponent(message));
    }

    @Shadow public abstract void teleportTo(double v, double v1, double v2);

    @Override
    public Level getLevel() {
        return level;
    }

    @Shadow public abstract double getX();
    @Shadow public abstract double getY();
    @Shadow public abstract double getZ();

    @Override
    public void teleport(systems.conduit.main.api.Entity entity) {
        this.teleport(entity.getX(), entity.getY(), entity.getZ());
    }

    @Override
    public void teleport(Position position) {
        this.teleport(position.x(), position.y(), position.z());
    }

    @Override
    public void teleport(double x, double y, double z) {
        this.teleportTo(x, y, z);
    }

    @Shadow public abstract boolean isSpectator();
    @Shadow public abstract void unRide();
    @Shadow public abstract EntityType<?> getType();
    @Shadow public abstract int getId();
    @Shadow public abstract void setId(int id);
    @Shadow public abstract Set<String> getTags();
    @Shadow public abstract boolean addTag(String tag);
    @Shadow public abstract boolean removeTag(String tag);
    @Shadow public abstract void kill();
    @Shadow public abstract Pose getPose();
    @Shadow public abstract int getPortalWaitTime();
    @Shadow public abstract void setSecondsOnFire(int time);
    @Shadow public abstract void setRemainingFireTicks(int ticks);
    @Shadow public abstract int getRemainingFireTicks();
    @Shadow public abstract void clearFire();
    @Shadow public abstract boolean isSilent();
    @Shadow public abstract void setSilent(boolean silent);
    @Shadow public abstract boolean isNoGravity();
    @Shadow public abstract void setNoGravity(boolean gravity);
    @Shadow public abstract boolean isInWater();
    @Shadow public abstract boolean isInWaterOrRain();
    @Shadow public abstract boolean isInWaterRainOrBubble();
    @Shadow public abstract boolean isInWaterOrBubble();
    @Shadow public abstract boolean isUnderWater();
    //@Shadow public abstract boolean isUnderLiquid(Tag<Fluid> fluid);
    @Shadow public abstract boolean isInLava();
    @Shadow public abstract void moveRelative(float scale, Vec3 vector);
    @Shadow public abstract void setLevel(Level level);
    @Shadow public abstract void moveTo(double x, double y, double z, float xRot, float yRot);
    @Shadow public abstract float distanceTo(Entity entity);
    @Shadow public abstract double distanceToSqr(double x, double y, double z);
    @Shadow public abstract double distanceToSqr(Entity entity);
    @Shadow public abstract double distanceToSqr(Vec3 vector);
    @Shadow public abstract Vec3 getEyePosition(float scale);
    @Shadow public abstract boolean isAlive();
    @Shadow public abstract boolean isInWall();
    @Shadow public abstract boolean startRiding(Entity entity);
    @Shadow public abstract void stopRiding();
    @Shadow public abstract boolean isSwimming();
    @Shadow public abstract void setSwimming(boolean state);
    @Shadow public abstract boolean isSprinting();
    @Shadow public abstract void setSprinting(boolean state);
    @Shadow public abstract boolean isGlowing();
    @Shadow public abstract void setGlowing(boolean state);
    @Shadow public abstract boolean isInvisible();
    @Shadow public abstract Team shadow$getTeam();
    @Shadow public abstract boolean isAlliedTo(Team team);
    @Shadow public abstract boolean isAlliedTo(Entity entity);
    @Shadow public abstract void setInvisible(boolean state);
    @Shadow public abstract int getMaxAirSupply();
    @Shadow public abstract void setAirSupply(int supply);
    @Shadow public abstract Direction getDirection();
    @Shadow public abstract Entity shadow$getControllingPassenger();
    @Shadow public abstract List<Entity> shadow$getPassengers();
    @Shadow public abstract boolean hasPassenger(Entity entity);
    @Shadow public abstract boolean hasPassenger(Class<? extends Entity> entity);
    @Shadow public abstract Collection<Entity> getIndirectPassengers();
    @Shadow public abstract boolean hasOnePlayerPassenger();
    @Shadow public abstract boolean isPassengerOfSameVehicle(Entity entity);
    //@Shadow public abstract boolean hasIndirectPassenger(Entity entity);
    @Shadow public abstract Vec3 position();

    @Shadow protected abstract SoundEvent shadow$getSwimSound();
    @Shadow protected abstract SoundEvent shadow$getSwimSplashSound();
    @Shadow protected abstract SoundEvent shadow$getSwimHighSpeedSplashSound();

    public Optional<Team> conduit_getTeam() {
        return Optional.of(this.shadow$getTeam());
    }

    public SoundEvent conduit_getSwimSound() {
        return shadow$getSwimSound();
    }

    public SoundEvent conduit_getSwimSplashSound() {
        return shadow$getSwimSplashSound();
    }

    public SoundEvent conduit_getSwimHighSpeedSplashSound() {
        return shadow$getSwimHighSpeedSplashSound();
    }

    public List<systems.conduit.main.api.Entity> conduit_getPassengers() {
        return this.shadow$getPassengers().stream().map(systems.conduit.main.api.Entity.class::cast).collect(Collectors.toList());
    }

    public Optional<systems.conduit.main.api.Entity> conduit_getControllingPassenger() {
        Entity entity = this.shadow$getControllingPassenger();
        if (entity == null) return Optional.empty();
        return Optional.of((systems.conduit.main.api.Entity) entity);
    }

    @Inject(method = "changeDimension", at = @At("HEAD"))
    public void changeDimension(ServerLevel level, CallbackInfoReturnable<Entity> callback) {
        EntityEvents.LevelSwitchEvent event = new EntityEvents.LevelSwitchEvent((Entity) (Object) this, (ServerLevel) this.getLevel(), level);
        Conduit.getEventManager().dispatchEvent(event);

        if (event.isCanceled()) return;
    }
}

File: src/main/kotlin/systems/conduit/core/mixins/api/EntityMixin.kt
package systems.conduit.core.mixins.api

import net.minecraft.core.Direction
import net.minecraft.core.Position
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.TextComponent
import net.minecraft.sounds.SoundEvent
import net.minecraft.tags.Tag
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.Pose
import net.minecraft.world.level.Level
import net.minecraft.world.level.material.Fluid
import net.minecraft.world.phys.Vec3
import net.minecraft.world.scores.Team
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow
import systems.conduit.api.Entity
import java.util.*
import java.util.stream.Collectors


@Mixin(value = [net.minecraft.world.entity.Entity::class], remap = false)
abstract class EntityMixin: Entity {

    @Shadow
    override var level: Level? = null

    @Shadow
    override var x = 0.0

    @Shadow
    override var y = 0.0

    @Shadow
    override var z = 0.0

    @Shadow
    abstract override fun getUUID(): UUID?

    @Shadow
    abstract override fun sendMessage(message: Component?)
    override fun sendMessage(message: String?) {
        sendMessage(TextComponent(message))
    }

    @Shadow
    abstract fun teleportTo(v: Double, v1: Double, v2: Double)

    override fun teleport(entity: Entity?) {
        this.teleport(entity!!.x, entity.y, entity.z)
    }

    override fun teleport(position: Position?) {
        this.teleport(position!!.x(), position.y(), position.z())
    }

    override fun teleport(x: Double, y: Double, z: Double) {
        teleportTo(x, y, z)
    }

    @Shadow
    abstract override fun isSpectator(): Boolean

    @Shadow
    abstract override fun unRide()

    @Shadow
    abstract override fun getType(): EntityType<*>?

    @Shadow
    abstract override fun getId(): Int

    @Shadow
    abstract override fun setId(id: Int)

    @Shadow
    abstract override fun getTags(): Set<String?>?

    @Shadow
    abstract override fun addTag(tag: String?): Boolean

    @Shadow
    abstract override fun removeTag(tag: String?): Boolean

    @Shadow
    abstract override fun kill()

    @Shadow
    abstract override fun getPose(): Pose?

    @Shadow
    abstract override fun getPortalWaitTime(): Int

    @Shadow
    abstract override fun setSecondsOnFire(time: Int)

    @Shadow
    abstract override fun setRemainingFireTicks(ticks: Int)

    @Shadow
    abstract override fun getRemainingFireTicks(): Int

    @Shadow
    abstract override fun clearFire()

    @Shadow
    abstract override fun isSilent(): Boolean

    @Shadow
    abstract override fun setSilent(silent: Boolean)

    @Shadow
    abstract override fun isNoGravity(): Boolean

    @Shadow
    abstract override fun setNoGravity(gravity: Boolean)

    @Shadow
    abstract override fun isInWater(): Boolean

    @Shadow
    abstract override fun isInWaterOrRain(): Boolean

    @Shadow
    abstract override fun isInWaterRainOrBubble(): Boolean

    @Shadow
    abstract override fun isInWaterOrBubble(): Boolean

    @Shadow
    abstract override fun isUnderWater(): Boolean

    @Shadow
    abstract override fun isUnderLiquid(fluid: Tag<Fluid?>?): Boolean

    @Shadow
    abstract override fun setInLava()

    @Shadow
    abstract override fun isInLava(): Boolean

    @Shadow
    abstract override fun moveRelative(scale: Float, vector: Vec3?)

    @Shadow
    abstract override fun moveTo(x: Double, y: Double, z: Double, xRot: Float, yRot: Float)

    @Shadow
    abstract fun distanceTo(entity: net.minecraft.world.entity.Entity?): Float

    @Shadow
    abstract override fun distanceToSqr(x: Double, y: Double, z: Double): Double

    @Shadow
    abstract fun distanceToSqr(entity: net.minecraft.world.entity.Entity?): Double

    @Shadow
    abstract override fun distanceToSqr(vector: Vec3?): Double

    @Shadow
    abstract override fun getEyePosition(scale: Float): Vec3?

    @Shadow
    abstract override fun isAlive(): Boolean

    @Shadow
    abstract override fun isInWall(): Boolean

    @Shadow
    abstract fun startRiding(entity: net.minecraft.world.entity.Entity?): Boolean

    @Shadow
    abstract override fun stopRiding()

    @Shadow
    abstract override fun isSwimming(): Boolean

    @Shadow
    abstract override fun setSwimming(state: Boolean)

    @Shadow
    abstract override fun isSprinting(): Boolean

    @Shadow
    abstract override fun setSprinting(state: Boolean)

    @Shadow
    abstract override fun isGlowing(): Boolean

    @Shadow
    abstract override fun setGlowing(state: Boolean)

    @Shadow
    abstract override fun isInvisible(): Boolean

    @Shadow
    abstract fun `shadow$getTeam`(): Team?

    @Shadow
    abstract override fun isAlliedTo(team: Team?): Boolean

    @Shadow
    abstract fun isAlliedTo(entity: net.minecraft.world.entity.Entity?): Boolean

    @Shadow
    abstract override fun setInvisible(state: Boolean)

    @Shadow
    abstract override fun getMaxAirSupply(): Int

    @Shadow
    abstract override fun setAirSupply(supply: Int)

    @Shadow
    abstract override fun getDirection(): Direction?

    @Shadow
    abstract fun `shadow$getControllingPassenger`(): net.minecraft.world.entity.Entity?

    @Shadow
    abstract fun `shadow$getPassengers`(): List<net.minecraft.world.entity.Entity?>

    @Shadow
    abstract fun hasPassenger(entity: net.minecraft.world.entity.Entity?): Boolean

    @Shadow
    abstract override fun hasPassenger(entity: Class<out net.minecraft.world.entity.Entity?>?): Boolean

    @Shadow
    abstract override fun getIndirectPassengers(): Collection<net.minecraft.world.entity.Entity?>?

    @Shadow
    abstract override fun hasOnePlayerPassenger(): Boolean

    @Shadow
    abstract fun isPassengerOfSameVehicle(entity: net.minecraft.world.entity.Entity?): Boolean

    @Shadow
    abstract fun hasIndirectPassenger(entity: net.minecraft.world.entity.Entity?): Boolean

    @Shadow
    abstract override fun position(): Vec3?

    @Shadow
    protected abstract fun `shadow$getSwimSound`(): SoundEvent?

    @Shadow
    protected abstract fun `shadow$getSwimSplashSound`(): SoundEvent?

    @Shadow
    protected abstract fun `shadow$getSwimHighSpeedSplashSound`(): SoundEvent?

    override fun conduit_getTeam(): Team? {
        return `shadow$getTeam`()
    }

    override fun conduit_getSwimSound(): SoundEvent? {
        return `shadow$getSwimSound`()
    }

    override fun conduit_getSwimSplashSound(): SoundEvent? {
        return `shadow$getSwimSplashSound`()
    }

    override fun conduit_getSwimHighSpeedSplashSound(): SoundEvent? {
        return `shadow$getSwimHighSpeedSplashSound`()
    }

    override fun conduit_getPassengers(): List<Entity?>? {
        return `shadow$getPassengers`().stream().map { obj: net.minecraft.world.entity.Entity? -> Entity::class.java.cast(obj) }.collect(Collectors.toList())
    }

    override fun conduit_getControllingPassenger(): Entity? {
        val entity = `shadow$getControllingPassenger`() ?: return null
        return entity as Entity
    }
}
